python/lsst/cp/pipe/utilsEfd.py

```diff
@@ -202,11 +206,67 @@ def selectTimeSeries(self, topicName, fields=[], startDate=None, endDate=None):
         table = Table(rows=series["values"], names=series["columns"], dtype=tableDtype)
         table["time"] = Time(table["time"], scale="utc")
         table.sort("time")
+        if 'private_sndStamp' in table.columns:
+            table['private_sndStamp'] = Time(table['private_sndStamp'], format='unix_tai')
+            table.sort("private_sndStamp")
 
         return table
 
+    def searchResults(self, data, dateStr, scale='tai'):
+        """Determine the entry for a specific date.
+
+        Parameters
+        ----------
+        data : `astropy.table.Table`
+            The table of results from the EFD.
+        dateStr : `str`
+            The date to look up in the status for.
+        scale : `str`, optional
+            Time scale to use.  Default is 'tai'.
+
+        Returns
+        -------
+        result = `astropy.table.Row`
+            The row of the data table corresponding to ``dateStr``.
+        """
+        dateValue = Time(dateStr, format='isot', scale=scale)
+        # Table is now sorted on "time", which is in UTC.
+
+        # Check that the date we want to consider is contained in the
+        # EFD data.
+        if data["time"][0] > dateValue or data["time"][-1] < dateValue:
+            raise RuntimeError("Requested date is outside of data range.")
+
+        # Binary search through the EFD entries in date, until the
+        # most recent monochromator state update prior to the spectrum
+        # in question is found.
+        low = 0
+        high = len(data)
+        idx = (high + low) // 2
+        found = False
+        iteration = 0
+        while not found:
+            if idx < 0 or idx > len(data) or iteration > 20:
+                raise RuntimeError("Search for date failed?")
```

Comment:
Question?  Also maybe put in the specific dateValue here.

Reply:
I don't think this can trigger outside of when I was debugging the search algorithm, but I've updated it to give information on what the state is when it failed.